extension/llm/src/function/create_embedding.cpp

```diff
@@ -15,31 +16,341 @@ using namespace kuzu::processor;
+            {"voyage-ai",
+                {{"voyage-3-large", 1024}, {"voyage-3.5", 1024}, {"voyage-3.5-lite", 1024},
+                    {"voyage-code-3", 1024}, {"voyage-finance-2", 1024}, {"voyage-law-2", 1024},
+                    {"voyage-code-2", 1536}}},
+            {"ollama", {{"nomic-embed-text", 768}, {"all-minilm:l6-v2", 384}}},
+            {"amazon-bedrock", {{"amazon.titan-embed-text-v1", 1024}}},
+            {"google-vertex", {{"gemini-embedding-001", 3072}, {"text-embedding-005", 768},
+                                  {"text-multilingual-embedding-002", 768}}},
+            {"google-gemini", {{"gemini-embedding-exp-03-07", 3072}, {"text-embedding-004", 768},
+                                  {"embedding-001", 768}}}};
+
+    auto providerItr = providerModelMap.find(provider);
+    if (providerItr == providerModelMap.end()) {
+        throw(BinderException("Invalid Provider: " + provider));
+    }
+    auto modelItr = providerItr->second.find(model);
+    if (modelItr == providerItr->second.end()) {
+        throw(BinderException("Invalid Model: " + model));
+    }
+    return modelItr->second;
+}
+
 static void execFunc(const std::vector<std::shared_ptr<common::ValueVector>>& parameters,
     const std::vector<common::SelectionVector*>& /*parameterSelVectors*/,
     common::ValueVector& result, common::SelectionVector* resultSelVector, void* /*dataPtr*/) {
-    // This iteration only supports using Ollama with nomic-embed-text.
-    // The user must install and have nomic-embed-text running at
-    // http://localhost::11434.
-    KU_ASSERT(parameters.size() == 1);
-    httplib::Client client("http://localhost:11434");
-    httplib::Headers headers = {{"Content-Type", "application/json"}};
+
+    auto provider = StringUtils::getLower(parameters[1]->getValue<ku_string_t>(0).getAsString());
+    auto model = StringUtils::getLower(parameters[2]->getValue<ku_string_t>(0).getAsString());
+    httplib::Client client(getClient(provider));
+    client.set_connection_timeout(30, 0);
+    client.set_read_timeout(30, 0);
+    client.set_write_timeout(30, 0);
+    httplib::Headers headers;
+
+    // Amazon Bedrock headers differ per request due to a signature authorization scheme
+    bool bedRock = provider == "amazon-bedrock";
+    if (!bedRock) {
+        headers = getHeaders(provider);
+    }
+    std::string path = getPath(provider, model);
+
     result.resetAuxiliaryBuffer();
     for (auto selectedPos = 0u; selectedPos < resultSelVector->getSelSize(); ++selectedPos) {
+
         auto text = parameters[0]->getValue<ku_string_t>(selectedPos).getAsString();
-        nlohmann::json payload = {{"model", "nomic-embed-text"}, {"prompt", text}};
-        auto res = client.Post("/api/embeddings", headers, payload.dump(), "application/json");
+
+        nlohmann::json payload = getPayload(provider, model, text);
+
+        if (bedRock) {
+            headers = getAWSBedrockHeader(payload);
```

Comment:
Hopefully this special casing can be removed by moving each provider behind a general abstraction.